### PR TITLE
MINOR: Upgrade Gradle wrapper version to 8.2.1

### PR DESCRIPTION
**Pull request by bmscomp**
Upgrade Gradle wrapper to the latest stable version 8.2.1, please refer to the related version's release notes 

https://docs.gradle.org/8.2.1/release-notes.html?_gl=1*d02o60*_ga*MTYzNzE5MjE3MS4xNjg5Mzc3MDgy*_ga_7W7NC6YNPT*MTY4OTYzNDIwNC4zLjEuMTY4OTYzNjcwNS4zOC4wLjA.

Notice from release notes : `We recommend users upgrade to 8.2.1 instead of 8.2. `

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by divijvaidya** [comment]
Need a change at https://github.com/apache/kafka/blob/2574bef61c7c53a8db6b23f8cde990cbe3fa2e11/gradlew#L118 as well. Please refer to your previous commit for the files that need to be changed. https://github.com/apache/kafka/commit/a7d0b3f753708a93aea92e614833f6f6e7443234

**Comment by bmscomp** [comment]
@divijvaidya  I already made the update, it's all good now there is three files to be changes 

```
dependencies.gradle
gradlew 
gradle-wrapper.properties 

```

Thanks for the remark

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,7 +1,7 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionSha256Sum=5625a0ae20fe000d9225d000b36909c7a0e0e8dda61c19b12da769add847c975
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.1.1-all.zip
+distributionSha256Sum=03ec176d388f2aa99defcadc3ac6adf8dd2bce5145a129659537c0874dea5ad1
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.1-bin.zip
```
Comment (divijvaidya) [inline gradle/wrapper/gradle-wrapper.properties]: we changed from "all" package to "bin". Is there a specific reason for this?

Comment (bmscomp) [reply gradle/wrapper/gradle-wrapper.properties]: @divijvaidya  Updated the pull request to use the all distribution

**Comment by divijvaidya** [comment]
Tests are failing with a related error.
```
* What went wrong:

Gradle could not start your build.

> Cannot create service of type BuildSessionActionExecutor using method LauncherServices$ToolingBuildSessionScopeServices.createActionExecutor() as there is a problem with parameter #21 of type FileSystemWatchingInformation.

   > Cannot create service of type BuildLifecycleAwareVirtualFileSystem using method VirtualFileSystemServices$GradleUserHomeServices.createVirtualFileSystem() as there is a problem with parameter #7 of type GlobalCacheLocations.

      > Cannot create service of type GlobalCacheLocations using method GradleUserHomeScopeServices.createGlobalCacheLocations() as there is a problem with parameter #1 of type List<GlobalCache>.

         > Could not create service of type FileAccessTimeJournal using GradleUserHomeScopeServices.createFileAccessTimeJournal().

            > Timeout waiting to lock journal cache (/home/jenkins/.gradle/caches/journal-1). It is currently in use by another Gradle instance.

              Owner PID: 4026719

              Our PID: 163366

              Owner Operation: 

              Our operation: 

              Lock file: /home/jenkins/.gradle/caches/journal-1/journal-1.lock
```

@bmscomp did you try running `./gradlew build` locally? Please ensure that build is successful for you.

**Comment by bmscomp** [comment]
@divijvaidya  It's a rebase issue, now that I rebased the branch into the trunk  as an issue is fixed before the execution is going on

**Review by divijvaidya — APPROVED** [review]
Unrelated test failures
```
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_20_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 8 and Scala 2.12 / kafka.network.SocketServerTest.processDisconnectedException()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/kafka.network/SocketServerTest/Build___JDK_8_and_Scala_2_12___processDisconnectedException__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders___2/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testSyncTopicConfigs()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_11_and_Scala_2_13___testSyncTopicConfigs__/)
[Build / JDK 11 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_11_and_Scala_2_13___testRackAwareRangeAssignor__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14032/7/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders__/)
```

Additionally, I have verified that [8.2 release notes](https://docs.gradle.org/8.2/release-notes.html) don't have any breaking changes and the SHA added to wrapper matches the SHA for download repository.

**Comment by divijvaidya** [comment]
We have had a large number of failures [1] since this was merged with `Timeout waiting to lock * cache (/home/jenkins/.gradle/caches/*/* It is currently in use by another Gradle instance.` 

Please investigate if you get a chance, else we may have to revert this commit tomorrow until we find the cause and fix.

[1] https://ge.apache.org/scans/failures?failures.failureMessage=Execution%20failed%20for%20task%20*%0A%3E%20Timeout%20waiting%20to%20lock%20*%20cache%20(/home/jenkins/.gradle/caches/*/*%20It%20is%20currently%20in%20use%20by%20another%20Gradle%20instance.%0A%20%20Owner%20PID:%20*%0A%20%20Our%20PID:%20*%0A%20%20Owner%20Operation:%20%0A%20%20Our%20operation:%20%0A%20%20Lock%20file:%20/home/jenkins/.gradle/caches/*/*/*&search.rootProjectNames=kafka&search.timeZoneId=Europe/Berlin

**Comment by bmscomp** [comment]
I could not  produce this behaviour in local machine, is there a way to run `./gradlew --stop ` before starting a build to stop all pending and running gradlew instances, this issue is not related specifically to this version but it can occurs in older versions 
 of gradle

**Comment by bmscomp** [comment]
@divijvaidya  I can see that all failures are related to `/home/jenkins/.gradle/caches/8.2.1/zinc-1.8.0_2.13.11_11/zinc-1.8.0_2.13.11_11.lock` dependency,  or the incremental scala compiler 


https://github.com/sbt/zinc/releases

The issue can be related to Gradle cache also !

**Comment by bmscomp** [comment]
@divijvaidya  In the `retry_zinc` file bash script in the kafka root directory there is something about this issue but related to an older version of gradle 

```

# Hacky workaround for https://github.com/gradle/gradle/issues/3777
# There is currently no configurable timeout, so we retry builds jenkins when we can't get a lock on the zinc compiler cache
# Hopefully we can remove this in the future, but this will save us from having to manually rebuild for the time being.
# Example:
# [2021-10-19T17:25:07.234Z] * What went wrong:
# [2021-10-19T17:25:07.234Z] Execution failed for task ':streams:streams-scala:compileScala'.
# [2021-10-19T17:25:07.234Z] > Timeout waiting to lock zinc-1.3.5_2.13.6_8 compiler cache (/home/jenkins/.gradle/caches/7.0.2/zinc-1.3.5_2.13.6_8). It is currently in use by another Gradle instance.
# [2021-10-19T17:25:07.234Z]   Owner PID: 3999
# [2021-10-19T17:25:07.234Z]   Our PID: 3973
# [2021-10-19T17:25:07.234Z]   Owner Operation: 
# [2021-10-19T17:25:07.234Z]   Our operation: 
# [2021-10-19T17:25:07.234Z]   Lock file: /home/jenkins/.gradle/caches/7.0.2/zinc-1.3.5_2.13.6_8/zinc-1.3.5_2.13.6_8.lock

```

Notice that we are using old version `1.8.0` of zinc incremental plugin and may be it will be a good idea to consider upgrading it toe a recent version `1.9.2`

https://github.com/sbt/zinc/releases

**Comment by divijvaidya** [comment]
Interestingly the failure rate has gone down today. I guess we will encounter it every time we upgrade gradle?! Let's wait and observer for another 24 hours.

Separately, upgrading Zinc is definitely a good idea.

**Comment by bmscomp** [comment]
@divijvaidya  Thank you so much for the report, yes the next time we upgrade Gradle version can cause those errors, I think it's better to clear cache, or stop all already running Gradle instances every version upgrade

**Pull request merged**